pkg/file/tar_index.go

```diff
@@ -0,0 +1,70 @@
+package file
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+type TarIndexVisitor func(TarIndexEntry) error
+
+// TarIndex is a tar reader capable of O(1) fetching of entry contents after the first read.
+type TarIndex struct {
+	indexByName map[string][]TarIndexEntry
+}
+
+// NewTarIndex creates a new TarIndex that is already indexed.
+func NewTarIndex(tarFilePath string, onIndex ...TarIndexVisitor) (*TarIndex, error) {
+	t := &TarIndex{
+		indexByName: make(map[string][]TarIndexEntry),
+	}
+	fh, err := os.Open(tarFilePath)
+	if err != nil {
+		return nil, err
+	}
+	defer fh.Close()
+	return t, t.indexEntries(fh, onIndex...)
+}
+
+// indexEntries records all tar header locations indexed by header names.
+func (t *TarIndex) indexEntries(file *os.File, onIndex ...TarIndexVisitor) error {
```

Comment:
It could be helpful to somehow indicate in the name of the `file` parameter that this is the TAR file, just to be clear that it's not a particular file contained in the TAR.